cmd/admin-handlers.go

```diff
@@ -601,46 +727,151 @@ func (a adminAPIHandlers) SetConfigHandler(w http.ResponseWriter, r *http.Reques
-	errs := writeTmpConfigPeers(globalAdminPeers, tmpFileName, configBytes)
+	peersErrs, err := a.setConfigDotJSON(ctx, configBytes)
+	if err != nil {
+		writeErrorResponseJSON(w, toAPIErrorCode(err), r.URL)
+		return
+	}
 
-	// Check if the operation succeeded in quorum or more nodes.
-	rErr := reduceWriteQuorumErrs(ctx, errs, nil, len(globalAdminPeers)/2+1)
-	if rErr != nil {
-		writeSetConfigResponse(w, globalAdminPeers, errs, false, r.URL)
+	if reduceWriteQuorumErrs(ctx, peersErrs, nil, len(globalAdminPeers)/2+1) != nil {
+		writeSetConfigResponse(w, globalAdminPeers, peersErrs, false, r.URL)
 		return
 	}
 
-	// Take a lock on minio/config.json. NB minio is a reserved
-	// bucket name and wouldn't conflict with normal object
-	// operations.
-	configLock := globalNSMutex.NewNSLock(minioReservedBucket, minioConfigFile)
-	if configLock.GetLock(globalObjectTimeout) != nil {
-		writeErrorResponseJSON(w, ErrOperationTimedOut, r.URL)
+	// serverMux (cmd/server-mux.go) implements graceful shutdown,
+	// where all listeners are closed and process restart/shutdown
+	// happens after 5s or completion of all ongoing http
+	// requests, whichever is earlier.
+	writeSetConfigResponse(w, globalAdminPeers, peersErrs, true, r.URL)
+
+	// Restart all node for the modified config to take effect.
+	sendServiceCmd(globalAdminPeers, serviceRestart)
+}
+
+func convertValueType(str string, jsonType gjson.Type) (interface{}, error) {
+	switch jsonType {
+	case gjson.False, gjson.True:
+		return strconv.ParseBool(str)
+	case gjson.JSON:
+		return gjson.Parse(str).Value(), nil
+	case gjson.String:
+		return str, nil
+	case gjson.Number:
+		return strconv.ParseFloat(str, 64)
+	default:
+		return nil, nil
+	}
+}
+
+// SetConfigKeysHandler - PUT /minio/admin/v1/config-keys
+func (a adminAPIHandlers) SetConfigKeysHandler(w http.ResponseWriter, r *http.Request) {
+
+	ctx := context.Background()
+	// Get current object layer instance.
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		writeErrorResponseJSON(w, ErrServerNotInitialized, r.URL)
 		return
 	}
-	defer configLock.Unlock()
 
-	// Rename the temporary config file to config.json
-	errs = commitConfigPeers(globalAdminPeers, tmpFileName)
-	rErr = reduceWriteQuorumErrs(ctx, errs, nil, len(globalAdminPeers)/2+1)
-	if rErr != nil {
-		writeSetConfigResponse(w, globalAdminPeers, errs, false, r.URL)
+	// Deny if WORM is enabled
+	if globalWORMEnabled {
+		writeErrorResponseJSON(w, ErrMethodNotAllowed, r.URL)
+		return
+	}
+
+	// Validate request signature.
+	adminAPIErr := checkAdminRequestAuthType(r, "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponseJSON(w, adminAPIErr, r.URL)
+		return
+	}
+
+	configBytes, err := a.getConfigDotJSON()
+	if err != nil {
+		writeErrorResponseJSON(w, toAPIErrorCode(err), r.URL)
+		return
+	}
+
+	origConfig := string(configBytes)
+	configStr := origConfig
+
+	queries := r.URL.Query()
+
+	for k, v := range queries {
+		var elem string
+		if len(v) > 0 {
```

Comment:
does this mean that the `value` cannot be a list?

Reply:
> does this mean that the value cannot be a list?

value is always a string, inside it we can have any type: string, int, json, list..

queries.Get(k) actually means v[0] literally according the Golang code: https://github.com/golang/go/blob/master/src/net/url/url.go#L810